src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs

```diff
@@ -548,6 +575,36 @@ public void StartProjectRequest(ICheckContext checkContext, string projectFullPa
         }
 
         private readonly Dictionary<int, List<BuildEventArgs>> _deferredEvalDiagnostics = new();
+
+        /// <summary>
+        /// Propagates a newly imported project file to all projects that import the original project file.
+        /// This method ensures that if Project A imports Project B, and Project B now imports Project C,
+        /// then Project A will also show Project C as an import.
+        /// </summary>
+        /// <param name="originalProjectFile">The path of the project file that is importing a new project.</param>
+        /// <param name="newImportedProjectFile">The path of the newly imported project file.</param>
+        private void PropagateImport(string originalProjectFile, string newImportedProjectFile)
```

Comment:
This all makes sense, though I was wondering whether we can benefit from following:
* We know that BuildEventArgs are received single-threaded
* We already have mapping of evaluationId --> project name (`_projectsByEvaluationId`)
* All `ProjectImportedEventArgs` of interest should have `BuildEventContext.EvaluationId` populated.

Apart from that - I'm wondering whether there might be API consuming cases where order of imports was important (e.g. a theoretical custom check validating that users do not try to import anything after custom sdk import targets - not to mess with the enforced targets)


That being said - can we consider `Dictionary<id, List<string>>` as datastructure for the defered data?

Reply:
As an extra points - we know the point in time when custom checks are done subscribing for a particualr project. At that point in time we should as well know if any check is subscribed to ProjectsImported - so we might be able to replay or discard deffered events for particular project at that point of time. And from that time on just directly feed the data to the checks (in case any subscribed).

But feel free to not solve as part of this PR if it'd turn more complicated than couple lines of code. (entering a lower priority item for that for future would be still nice though)